src/sentry/rules/actions/notify_email.py

```diff
@@ -0,0 +1,518 @@
+"""
+Used for notifying via email
+"""
+from __future__ import absolute_import
+
+import itertools
+import logging
+import six
+
+from django import forms
+from django.utils import dateformat
+from django.utils.encoding import force_text
+from django.utils.safestring import mark_safe
+
+from sentry.db.models.fields.bounded import BoundedBigIntegerField
+from sentry.digests.utilities import get_digest_metadata, get_personalized_digests
+from sentry.models import ProjectOwnership, User, Team, Project
+from sentry.plugins.base.structs import Notification
+from sentry.plugins.sentry_mail.models import MailPlugin
+from sentry.rules.actions.base import EventAction
+from sentry.utils import metrics
+from sentry.utils.cache import cache
+from sentry.utils.committers import get_serialized_event_file_committers
+from sentry.utils.email import MessageBuilder, group_id_to_email
+from sentry.utils.linksign import generate_signed_link
+
+from enum import Enum
+
+
+class NotifyEmailActionTargetType(Enum):
+    ISSUE_OWNERS = "IssueOwners"
+    TEAM = "Team"
+    MEMBER = "Member"
+
+
+CHOICES = [
+    (NotifyEmailActionTargetType.ISSUE_OWNERS.value, "Issue Owners"),
+    (NotifyEmailActionTargetType.TEAM.value, "Team"),
+    (NotifyEmailActionTargetType.MEMBER.value, "Member"),
+]
+
+
+class MailAdapter(object):
+    """
+    Adapter for the `MailPlugin`. This class attempts to decouple `MailPlugin` from the email notification question.
+    On complete decoupling, consider renaming this class to `MailProcessor`
+
+    Logic that was used solely for sending mail notifications is extracted into this class.
+
+    We want the mail behaviours in the `MailAdapter` to be as similar as possible to the `MailPlugin`.
+    Thus, keys are still fetched from the `MailPlugin` as they are used to access project options for mail.
+    Essentially, we still treat MailPlugin as a source of truth for keys.
+    """
+
+    project_default_enabled = True
+    project_conf_form = None
+    logger = logging.getLogger("sentry.notify_email_action.MailAdapter")
+    legacy_mail = MailPlugin()
+
+    def rule_notify(self, event, futures, target_type, target_identifier=None):
```

Comment:
It would be easier to extend the original class to support handling more parameters. By that, what I mean is using `unsplit_key_for_targeted_action` and providing more args to `self.notify(notification, target_type, target_identifier)`.

I initially understood your motivation for extracting and duplicating code would be to gradually decouple dependencies from the original class-- to the point that there is a manageable number of shared logic that we can create a util class for.

Reply:
I think ideally we'd extract shared logic out and have it used by both places rather than duplicating. The duplication works ok since it's just temporary, but it makes it a bit harder to review